pkg/agent/controller/l7visibility/l7visibility_controller.go

```diff
@@ -0,0 +1,404 @@
+	}
+	l7c.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    l7c.addPod,
+			UpdateFunc: l7c.updatePod,
+			DeleteFunc: l7c.deletePod,
+		},
+		resyncPeriod,
+	)
+	l7c.namespaceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    l7c.addNamespace,
+			UpdateFunc: l7c.updateNamespace,
+		},
+		resyncPeriod,
+	)
+	return l7c
+}
+
+func (l7c *L7VisibilityController) Run(stopCh <-chan struct{}) {
+	defer l7c.queue.ShutDown()
+	klog.InfoS("Starting", "Controller", controllerName)
+	defer klog.InfoS("Shutting down", "Controller", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, l7c.podListerSynced, l7c.namespaceListerSynced) {
+		return
+	}
+	if itf, ok := l7c.interfaceStore.GetInterfaceByName(config.L7NetworkPolicyTargetPortName); ok {
+		l7c.targetPort = uint32(itf.OFPort)
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(l7c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (l7c *L7VisibilityController) worker() {
+	for l7c.processNextWorkItem() {
+	}
+}
+
+func (l7c *L7VisibilityController) processNextWorkItem() bool {
+	obj, quit := l7c.queue.Get()
+	if quit {
+		return false
+	}
+	defer l7c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		// As the item in the work queue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		l7c.queue.Forget(key)
+		klog.ErrorS(nil, "Expected string in work queue but got", "key", obj)
+		return true
+	} else if err := l7c.syncPod(key); err == nil {
+		// If no error occurs we Forget this item, so it does not get queued again until
+		// another change happens.
+		l7c.queue.Forget(key)
+	} else if err == errInvalidAnnotation || err == errPodInterfaceNotFound {
+		// Handle errors
+		l7c.queue.Forget(key)
+	} else {
+		// Put the item back on the work queue to handle any transient errors.
+		l7c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing Pod object for L7Visibility failed, requeue", "Pod", key)
+	}
+	return true
+}
+
+func (l7c *L7VisibilityController) addPod(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	if !isValidPod(pod) {
+		return
+	}
+	_, ok := pod.Annotations[types.L7VisibilityAnnotationKey]
+	if !ok {
+		if !l7c.namespaceAnnotationExists(pod) {
+			return
+		}
+	}
+	klog.V(2).InfoS("Processing Pod ADD event", "Pod", klog.KObj(pod))
+	podNN := k8s.NamespacedName(pod.Namespace, pod.Name)
+	l7c.queue.Add(podNN)
+}
+
+func (l7c *L7VisibilityController) updatePod(oldObj interface{}, obj interface{}) {
+	oldPod := oldObj.(*v1.Pod)
+	updatedPod := obj.(*v1.Pod)
+	if !isValidPod(updatedPod) {
+		return
+	}
+	oldAnnotation, _ := oldPod.GetAnnotations()[types.L7VisibilityAnnotationKey]
+	updatedAnnotation, updatedAnnotationOk := updatedPod.GetAnnotations()[types.L7VisibilityAnnotationKey]
```

Comment:
why do you use `GetAnnotations()` here instead of just accessing the `.Annotations` field? This doesn't seem consistent with the rest of the code.